### PR TITLE
Support ragged gather reduce kernel fan-in in forward pass

### PR DESCRIPTION
**Pull request by NuojCheng**
# Description

Before this PR, the fan-in feature of `ragged_gather_reduce` kernel is not properly used in the forward pass. We still have weight sum after gather reduce kernel and no fan in is used. This PR corrects this mistake and further improve performance.

# Tests

* Step time before change: [19.05 s](http://xprof.corp.google.com/trace_viewer/chengnuojin-1068214515790825753)
* Step time after change: [17.8 s](http://xprof.corp.google.com/trace_viewer/chengnuojin-9288426616673894265)

Performance gets improved 7%.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/moe.py [hunk 1]
```diff
@@ -839,38 +839,44 @@ def unpermute(
...
-        reshaped_weights = jnp.ones_like(reshaped_weights)
-      if self.config.float32_weight_sum:
-        reshaped_intermediate = reshaped_intermediate.astype(jnp.float32)
-        reshaped_weights = reshaped_weights.astype(jnp.float32)
-      output = jnp.einsum(
-          "BKE,BK -> BE",
-          reshaped_intermediate,
-          reshaped_weights,
-          precision=matmul_precision,
+      reshaped_weights = jnp.reshape(weights, (-1, self.num_experts_per_tok))
```
Comment (RissyRan) [inline src/maxtext/layers/moe.py:861]: nit: are those format change? if needed?

File: src/maxtext/kernels/ragged/ragged_sort.py [hunk 2]
```diff
@@ -213,44 +228,48 @@ def _ring_ragged_unsort_fwd(sorted_tokens_local, group_sizes_local, topk_argsort
...
     )
-    return grad_sorted_tokens, None, None
+    return grad_sorted_tokens, None, None, None
 
   _ring_ragged_unsort.defvjp(_ring_ragged_unsort_fwd, _ring_ragged_unsort_bwd)
 
-  return _ring_ragged_unsort(sorted_tokens_local, group_sizes_local, topk_argsort_revert_indices)
+  # Build the flat weights array from the routing weights.
+  n = topk_argsort_revert_indices.shape[0]
+  if topk_weights is None:
```
Comment (RissyRan) [inline src/maxtext/kernels/ragged/ragged_sort.py]: Wondering when it will be None? Could you help add a comment?

Or should we expect this to be non-None for perf reason?

Comment (NuojCheng) [reply src/maxtext/kernels/ragged/ragged_sort.py]: `topk_weights` is an optional input, when not assigned, it assumes equal weights. I added comments in the funciton docstring. Good point!

File: src/maxtext/layers/moe.py (hunk 1 above)
Comment (NuojCheng) [reply src/maxtext/layers/moe.py:861]: Yes this is just format change. weight_sum in only needed when ragged gather reduce is not used.

File: src/maxtext/kernels/ragged/ragged_sort.py [hunk 3]
```diff
@@ -143,39 +143,55 @@ def _ring_ragged_sort_bwd(res, g_out):
...
   Backward:
-    ``g_sorted_tokens[j] = g_out[i]`` where ``j = topk_argsort_revert_indices[i]``
-    and ``j`` is in ``[shard_output_start, shard_output_end)``. Since the source indices
-    represent a permutation, this is a masked ragged gather over the inverse permutation.
+    ``g_sorted_tokens[j] = w[i] * g_out[i // topk]`` where
+    ``j = topk_argsort_revert_indices[i]`` and ``j`` is in
+    ``[shard_output_start, shard_output_end)``.
 
   Args:
     sorted_tokens_local: 2D ``[num_tokens_local * topk, hidden]`` output tensor from local experts.
```
Comment (gobbleturk) [inline src/maxtext/kernels/ragged/ragged_sort.py:171]: for my own understandming num_tokens_local * topk is not really a multiplication - its really just the number of local tokens after the topk blowup (e.g. tokens have been duplicated as necessary locally) - in general it will be less than a factor of topk - each token only belongs to a few local expert (different tokens  have different local counts?)

**Review by gobbleturk — APPROVED** [review]

File: src/maxtext/kernels/ragged/ragged_sort.py (hunk 3 above)
Comment (NuojCheng) [reply src/maxtext/kernels/ragged/ragged_sort.py:171]: I am using JAX annotation (ragged_sort.py still in JAX world) so the `sorted_tokens_local` is still with blow-up shape I think.

File: src/maxtext/layers/moe.py (hunk 1 above)
Comment (RissyRan) [reply src/maxtext/layers/moe.py:861]: OK, I meant to say, we could skip this change if main branch is not affected by this (i.e. failure).

Comment (NuojCheng) [reply src/maxtext/layers/moe.py:861]: I did not change the code but only added indentation. Somehow github diff does not properly reflect this...

File: src/maxtext/kernels/ragged/ragged_sort.py (hunk 2 above)
Comment (NuojCheng) [reply src/maxtext/kernels/ragged/ragged_sort.py]: Removed

**Review by RissyRan — APPROVED** [review]
